public/app/core/context/ModalsContextProvider.tsx

```diff
@@ -0,0 +1,114 @@
+import React, { useEffect, useState } from 'react';
+
+import { textUtil } from '@grafana/data';
+import { locationService } from '@grafana/runtime';
+import { ConfirmModal, ConfirmModalProps, ModalsContext } from '@grafana/ui';
+import { ModalsContextState } from '@grafana/ui/src/components/Modal/ModalsContext';
+import { ShowConfirmModalEvent, ShowModalReactEvent } from 'app/types/events';
+
+import appEvents from '../app_events';
+
+export interface Props {
+  children: React.ReactNode;
+}
+
+/**
+ * Implements the ModalsContext state logic (not used that much, only needed in edge cases)
+ * Also implements the handling of the events ShowModalReactEvent and ShowConfirmModalEvent.
+ */
+export function ModalsContextProvider(props: Props) {
+  const [state, setState] = useState<ModalsContextState>({
+    component: null,
+    props: {},
+    showModal: (component: React.ComponentType<any>, props: any) => {
+      setState({ ...state, component, props });
+    },
+    hideModal: () => {
+      setState({ ...state, component: null, props: {} });
+    },
+  });
+
+  useEffect(() => {
```

Comment:
I'm not sure how much it matters in practice, but should this `useEffect` return a clean up method with the following?
```
() => appEvents.removeAllListeners();
```

Reply:
@tomratcliffe technically it should, but since it's a "singleton" top-level context that will only exist once and be unmounted when app is unmounted it does not really matter